reference/preconditioner/isai_kernels.cpp

```diff
@@ -0,0 +1,217 @@
+
+    for (size_type row = 0; row < size[0]; ++row) {
+        const auto i_row_begin = i_row_ptrs[row];
+        const auto i_row_end = i_row_ptrs[row + 1];
+        const auto i_row_elems = i_row_end - i_row_begin;
+
+        trisystem_array.resize_and_reset(i_row_elems * i_row_elems);
+        auto trisystem = trisystem_array.get_data();
+        std::fill_n(trisystem, i_row_elems * i_row_elems, zero<ValueType>());
+
+        for (size_type i = 0; i < i_row_elems; ++i) {
+            const auto col = i_cols[i_row_begin + i];
+            const auto m_row_end = m_row_ptrs[col + 1];
+            auto m_row_ptr = m_row_ptrs[col];
+            auto i_row_ptr = i_row_begin;
+            // Values in `trisystem` will be stored in transposed order
+            // to prevent the need for an explicit transpose.
+            // Since `trisystem` stores in column major, this is equivalent
+            // to storing it in row-major and reading in column major.
+            auto idx = i * i_row_elems;
+            while (m_row_ptr < m_row_end && i_row_ptr < i_row_end) {
+                const auto sparsity_col = i_cols[i_row_ptr];
+                const auto m_col = m_cols[m_row_ptr];
+                if (sparsity_col == m_col) {
+                    trisystem[idx] = m_vals[m_row_ptr];
+                    ++m_row_ptr;
+                    ++i_row_ptr;
+                    ++idx;
+                } else if (m_col < sparsity_col) {
+                    ++m_row_ptr;
+                } else {  // element not present -> leave value at 0
+                    ++i_row_ptr;
+                    ++idx;
+                }
+            }
+        }
+
+        rhs_array.resize_and_reset(i_row_elems);
+        auto rhs = rhs_array.get_data();
+
+        trs_solve(i_row_elems, trisystem, rhs);
+
+        // Drop RHS as a row to memory (since that is the computed inverse)
+        std::copy_n(rhs, i_row_elems, i_vals + i_row_begin);
+    }
+
+    // Call make_srow
+    matrix::CsrBuilder<ValueType, IndexType>{inverse_mtx};
+}
+
+
+template <typename ValueType, typename IndexType>
+void generate_l(std::shared_ptr<const DefaultExecutor> exec,
+                const matrix::Csr<ValueType, IndexType> *l_csr,
+                matrix::Csr<ValueType, IndexType> *inverse_l)
+{
+    auto trs_solve = [](IndexType size, ValueType *trisystem, ValueType *rhs) {
+        if (size <= 0) {
+            return;
+        }
+        // RHS is the identity: zero everywhere except for the last entry
+        // since that is the row we are searching the inverse for
+        std::fill_n(rhs, size - 1, zero<ValueType>());
+        rhs[size - 1] = one<ValueType>();
+
+        // Note: `trisystem` is an upper triangular matrix here (stored in
+        //       colum major)
+        for (IndexType d_col = size - 1; d_col >= 0; --d_col) {
+            const auto diag = trisystem[d_col * size + d_col];
+            const auto bot = rhs[d_col] / diag;
```

Comment:
What happens if `diag` is zero (bot = nan) or close to zero (badly conditioned) ?

Reply:
If `diag` is zero, then bot is actually `+/-infinity`. However, I don't really see any way around it in this case. I don't think I can do any kind of pivoting in a simple triangular solve.
Unfortunately, we don't have any error-channel, so we can't directly tell the user that something went wrong. It might be better to have `infinity` or `NaN` in the matrix, so the problem can be traced back.